### ISSUE TITLE
Moq still doesn't support multithreading

### ISSUE THREAD

**Issue by VitaliyKurokhtin** [moq/moq4#91]
Just came across following issue with Moq (4.2.1312.1622). Consider interface:

```csharp
public IMyInterface
{
    void Do(int arg);
}
```

... setup:

```csharp
var mock = Mock.Create<IMyInterface>();
mock.Setup(x => x.Do(0)).Verifiable();
mock.Setup(x => x.Do(1)).Verifiable();
```

... and invocation:

```csharp
var obj = mock.Object;
var t1 = Task.Run(() => obj.Do(0));
var t2 = Task.Run(() => obj.Do(1));

Task.WhenAll(t1, t2).Wait();
mock.Verify();
```

Every now and then this verification will fail saying that one of two setups was not matched. But will succeed if changed to:

```csharp
mock.Verify(x => x.Do(0), Times.Once);
mock.Verify(x => x.Do(1), Times.Once);
```

This happens because method call invocation logic inside Interceptor is not thread safe. When Interceptor.Intercept method is called it creates strategy pipeline, one of strategies in this pipeline - ExtractProxyCall - locates matched method call and places it into InterceptStrategyContext.CurrentCall. Which is later used by ExecuteCall strategy to do the actual call.

The problem here is that both executing threads will work with same InterceptorContext instance. So when executed in parallel it's possible of one ExtractProxyCall strategy to overwrite method call just located by another ExtractProxyCall strategy, that way both ExecuteCall strategies will end up executing same method call twice. This can easily be confirmed by attaching callbacks to original setups.

mock.Verify(x => ...) on the other hand does verification evaluating given expression against collected call contexts which, obviously, contains correct information.

**Comment by stakx** [comment]
@VitaliyKurokhtin: I could not reproduce it even after running your test approx. 100 times. Could you please update to the latest version of Moq and see if this issue persists?

**Comment by stakx** [comment]
@VitaliyKurokhtin: I am closing this issue as I strongly suspect that this has been fixed long ago (possibly by commit 19485087a436d95643eb12917574e72d8fbd70b0). Most of what you mentioned above no longer exists today in the same form. For instance, `InterceptStrategyContext.CurrentCall` appears to have been refactored into `CurrentInterceptorContext.Call`, and each interception works with a different copy of that.

I tried reproducing this issue (but didn't succeed).
